### PR TITLE
[0101] - Bulk updated applications_open_from to '13/10/20'

### PR DESCRIPTION
**Pull request by avinhurry**
### Context

Providers are struggling to publish courses due to the application_open_from date being in the past.  When they update the course it throws a validation error (which is not surfaced on the front end).

### Changes proposed in this pull request

Updated applications_open_from date for the affected 17390 courses.

### Guidance to review

Restore backup database from production and run both API and Publish.

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by avinhurry** [comment]
[Wrong date stated for Application open date for courses rolled over on Publish](https://trello.com/c/0ewZlfww/101-wrong-date-stated-for-application-open-date-for-courses-rolled-over-on-publish)

File: db/migrate/20200910143249_fix_course_application_open_from_dates.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class FixCourseApplicationOpenFromDates < ActiveRecord::Migration[6.0]
+  def up
+    recruitment_cycle = RecruitmentCycle.find_by(year: "2021")
+    return if recruitment_cycle.nil?
+
+    courses = recruitment_cycle.courses.where.not(:applications_open_from => recruitment_cycle.application_start_date..recruitment_cycle.application_end_date)
```
Comment (darokel) [inline db/migrate/20200910143249_fix_course_application_open_from_dates.rb:6]:
```suggestion
    courses = recruitment_cycle.courses.where.not(applications_open_from: recruitment_cycle.application_start_date..recruitment_cycle.application_end_date)
```

File: db/migrate/20200910143249_fix_course_application_open_from_dates.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+    courses.update_all applications_open_from: recruitment_cycle.application_start_date
```
Comment (darokel) [inline db/migrate/20200910143249_fix_course_application_open_from_dates.rb:8]:
```suggestion
courses.update_all(applications_open_from: recruitment_cycle.application_start_date)
```

File: db/migrate/20200910143249_fix_course_application_open_from_dates.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+class FixCourseApplicationOpenFromDates < ActiveRecord::Migration[6.0]
+  def up
+    recruitment_cycle = RecruitmentCycle.find_by(year: "2021")
+    return if recruitment_cycle.nil?
+
+    courses = recruitment_cycle.courses.where.not(applications_open_from: recruitment_cycle.application_start_date..recruitment_cycle.application_end_date)
```
Comment (gpeng) [inline db/migrate/20200910143249_fix_course_application_open_from_dates.rb:6]: Can we split this line up a bit...

```suggestion
   valid_start_date_range = recruitment_cycle.application_start_date..recruitment_cycle.application_end_date
   courses = recruitment_cycle.courses.where.not(applications_open_from: valid_start_date_range)
```

File: db/migrate/20200910143249_fix_course_application_open_from_dates.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+    courses.update_all(applications_open_from: recruitment_cycle.application_start_date)
+  end
+
+  def down
+    # cannot go back as this will be invalid
```
Comment (gpeng) [inline db/migrate/20200910143249_fix_course_application_open_from_dates.rb:12]: We can make this reversible if we want to be more cautious. We could add a new column e.g. `old_applications_from_value` and copy the existing values to it before the update. Then drop the column in a subsequent migration in a future PR once we're happy this one didn't cause issues. Whether we do that depends on how happy we are with the risk. Might be overkill here 🤷

**Review by gpeng — COMMENTED** [review]
I'm still a bit against using ActiveRecord in migrations as the domain objects change and as migrations don't break tests so they get forgotten and don't work in the future. I'm also not too keen on updating the data here but if we do maybe we should use SQL which is a more appropriate language in this layer (the DB layer essentially). 

Maybe using something like [after_party](https://github.com/theSteveMitchell/after_party) might be something to look at?

File: db/migrate/20200910143249_fix_course_application_open_from_dates.rb (hunk 4 above)
Comment (gpeng) [reply db/migrate/20200910143249_fix_course_application_open_from_dates.rb:14]: Discussed offline. We're going to proceed as is with this and review carefully in staging.

**Review by darokel — APPROVED** [review]
Assuming that this has been discussed offline with @gpeng  👍

**Review by gpeng — APPROVED** [review]

**Pull request merged**